webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,151 @@
+import { forEach, reduce, trimEnd } from 'lodash';
+
+import { decode_keys, KeyCache } from './keyCache';
+import { StreamSpec } from './projects';
+
+import { Streamable, StreamEntityMap } from '.';
+
+// About 60 seconds of auto-retry.
+const backoffs = [0, 1, 2, 4, 8, 10, 10, 10, 15];
+const sleep = (s: number) => new Promise((r) => setTimeout(r, 1000 * s));
+
+type Entity = {
+    keyCache: KeyCache,
+    spec: StreamSpec
+}
+
+export class Stream {
+
+    #ws: WebSocket;
+    #retries: number = 0;
+    #numSyncs: number = 0;
+
+    #entities: Array<Record<Streamable, Entity>> = [];
+    #curEntity: Record<Streamable, Entity> | undefined;
```

Comment:
Entity is a weird name for this... maybe "current subscription"?  Especially because a single subscription for the streaming client can include subscription specs for multiple entities.  Or... at least that will be possible when we support multiple streamable entitites...